src/dbnode/x/xio/segment_reader.go

```diff
@@ -48,28 +50,29 @@ func (sr *segmentReader) Read(b []byte) (int, error) {
 	if len(b) == 0 {
 		return 0, nil
 	}
-	var head, tail []byte
-	if b := sr.segment.Head; b != nil {
-		head = b.Bytes()
+
```

Comment:
I think you can avoid this whole use of buffers by doing it something like this, what do you think?

```golang
type segmentReader struct {
 size int
 headSet bool
 tailSet bool
 segment  ts.Segment
}

func (sr *segmentReader) Read(b []byte) (int, error) {
 if len(b) == 0 {
  return 0, nil
 }

 if sr.headSet && sr.tailSet {
  return 0, io.EOF
 }

 if !sr.headSet {
  sr.headSet = true
  if h := sr.segment.Head; h != nil {
   n := copy(b, h.Bytes()) 
   return n, nil
  }
 }
 
 sr.tailSet = true
 if t := sr.segment.Tail; t != nil {
  n := copy(b, t.Bytes()) 
  return n, nil
 }

 return 0, io.EOF
}
```

Reply:
Hmm yeah so I definitely agree that approach is simpler. I think it is slightly different logic-wise, though, in that the existing implementation seems to support "partial" segments whereas the above doesn't. 

As-is, it supports the case where you first read to the end of the segment, but then if more bytes were added to that underlying buffer, a subsequent read would returns those additional bytes. Whereas the above is one-time-use in that a segment_reader will read to the end only once regardless of if new bytes are added. So the question is (a) am I understanding the diff here correctly? and (b) assuming yes then is it important we preserve that behavior or can we assume immutable segment heads/tails?

Reply:
So as far as I can tell, if you do any kind of meddling with the buffer afterwards, this whole model falls apart (e.g. if you had a head of length 10 and a tail of length 5, called read twice to get to the end, then changed head to be length 15, you'd double read the tail rather than do the correct thing)

In either case, the `lazyXX` approach would also hit the same issue, since it won't be updated if the underlying segment changes.

I'll loop in the other guys to see if there was a motivation for it being done this way previously

Reply:
Yeah all very good points. For the `lazyXX` I was overlooking the the underlying buffer ref we take when we call `.Bytes()` wouldn't actually reflect any changes to the checked bytes if new bytes were appended to them.

So it probably is just unnecessary complexity in the current implementation that we can remove. Agreed we can confirm w/ others though before we merge.

Reply:
Pushed w/ your implementation. Only modification I made was to handle the case where you handle reading head and tail in a single `Read` call (e.g. you have 5 head and 5 tail, and a buffer of 8, so Read should take 5 from head and 3 from tail in that single call). Noticed this from a unit test case.

Reply:
Ok so actually another set of test cases failing w/ this new approach is where calls to `Read(b)` are passing in a buffer `b` whose length is < the head or tail length. And so to deplete the segment multiple `Read` calls are made. W/ our changes we are expecting to be able to copy the whole head or tail in a single call.

So I actually am thinking the original change where we just cache the `[]byte` from the `checked.Bytes()` call is the best we can do assuming we want to preserve the reader's behavior of allowing incremental calls to `Read`.

Reply:
Makes sense, gotcha